fawltydeps/types.py

```diff
@@ -47,7 +48,25 @@ def __str__(self) -> str:
 
 
 @dataclass(frozen=True, eq=True, order=True)
-class CodeSource:
+class Source(ABC):
+    """Base class for some source of input to FawltyDeps.
+
+    This exists to inject the class name of the subclass into our JSON output.
+    """
+
+    source_type: Type["Source"] = field(init=False)
+
+    def __post_init__(self) -> None:
+        object.__setattr__(self, "source_type", self.__class__)
+
+    @abstractmethod
+    def render(self, detailed: bool) -> str:
+        """Return a human-readable string representation of this source."""
+        raise NotImplementedError
```

Comment:
Why is this an `abstractmethod`?

Reply:
To force/require subclasses to implement it. In practice, I believe the `@abstractmethod` provides for a better error message if we somehow end up calling it at runtime.